lib/Dialect/Polynomial/Conversions/PolynomialToModArith/PolynomialToModArith.cpp

```diff
@@ -1268,6 +1268,190 @@ struct ConvertINTT : public OpConversionPattern<INTTOp> {
+    Value x = op.getOperand();
+    auto terms = poly.getTerms();
+
+    if (terms.empty()) {
+      auto zeroConst = b.create<arith::ConstantOp>(
+          b.getI64Type(), b.getIntegerAttr(b.getI64Type(), 0));
+      rewriter.replaceOp(op, zeroConst);
+      return success();
+    }
+
+    // Get the highest degree
+    int64_t maxDegree = terms.back().getExponent().getSExtValue();
+    const int DEGREE_THRESHOLD = 5;
+    if (maxDegree < DEGREE_THRESHOLD) {
+      return failure();  // Let other patterns handle low-degree polynomials
+    }
+
+    // Create a map from exponent to coefficient
+    std::map<int64_t, int64_t> coeffMap;
+    for (auto term : terms) {
+      coeffMap[term.getExponent().getSExtValue()] =
+          term.getCoefficient().getSExtValue();
+    }
+
+    // Choose k optimally - sqrt of maxDegree is typically a good choice
+    int64_t k = std::max(static_cast<int64_t>(std::ceil(std::sqrt(maxDegree))),
+                         static_cast<int64_t>(1));
+
+    // Precompute x^1, x^2, ..., x^k
+    std::vector<Value> xPowers(k + 1);
+    xPowers[0] = b.create<arith::ConstantOp>(
+        b.getI64Type(), b.getIntegerAttr(b.getI64Type(), 1));
+    xPowers[1] = x;
+    for (int64_t i = 2; i <= k; i++) {
+      xPowers[i] = b.create<arith::MulIOp>(xPowers[i - 1], x);
+    }
+
+    // Number of chunks we'll need
+    int64_t m = (maxDegree + k - 1) / k;  // ceiling division
+
+    // Initialize array for chunk polynomials
+    std::vector<Value> chunkValues(m, nullptr);
+
+    // Evaluate each chunk using Horner's method
+    for (int64_t i = 0; i < m; i++) {
+      // Start with coefficient of degree (i+1)*k-1, if present
+      int64_t highestDegreeInChunk = std::min((i + 1) * k - 1, maxDegree);
+      int64_t lowestDegreeInChunk = i * k;
+
+      Value chunkValue = nullptr;
+      bool hasCoeff = false;
+
+      // Apply Horner's method within the chunk
+      for (int64_t j = highestDegreeInChunk; j >= lowestDegreeInChunk; j--) {
+        if (coeffMap.count(j)) {
+          if (!hasCoeff) {
+            // First coefficient in this chunk
+            chunkValue = b.create<arith::ConstantOp>(
+                b.getI64Type(), b.getIntegerAttr(b.getI64Type(), coeffMap[j]));
+            hasCoeff = true;
+          } else {
+            // Apply Horner's step: multiply by x and add next coefficient
+            chunkValue = b.create<arith::MulIOp>(chunkValue, x);
+            auto coeff = b.create<arith::ConstantOp>(
+                b.getI64Type(), b.getIntegerAttr(b.getI64Type(), coeffMap[j]));
+            chunkValue = b.create<arith::AddIOp>(chunkValue, coeff);
+          }
+        } else if (hasCoeff) {
+          // No coefficient for this degree, but we have previous coefficients
+          chunkValue = b.create<arith::MulIOp>(chunkValue, x);
+        }
+      }
+
+      if (hasCoeff) {
+        chunkValues[i] = chunkValue;
+      }
+    }
+
+    // Combine chunks using Horner's method with x^k
```

Comment:
This is a nice idea! Vs precomputing powers x^{tk} it saves k=sqrt(n) plaintext-ciphertext ops.